### PR TITLE
Initial implementation of generic dictionary access for R2R (Supports TypeHandles on Windows x64 only)

### PR DESCRIPTION
**Pull request by fadimounir**
@jkotas, I'd like to get some initial feedback on this version of the dictionary access implementation (incomplete work... just works for a simple ldtoken instruction on Windows)

### CODE REVIEW

File: src/vm/zapsig.cpp [hunk 1]
```diff
@@ -786,14 +787,18 @@ TypeHandle ZapSig::DecodeType(Module *pEncodeModuleContext,
...
+
     return th;
 }
 
 MethodDesc *ZapSig::DecodeMethod(Module *pReferencingModule,
                                  Module *pInfoModule,
                                  PCCOR_SIGNATURE pBuffer,
-                                 TypeHandle * ppTH /*=NULL*/)
+                                 TypeHandle * ppTH, /*=NULL*/
+                                 PCCOR_SIGNATURE* ppRemainingSig /*=NULL*/)
```
Comment (jkotas) [inline src/vm/zapsig.cpp]: The `ppRemainingSig` should not be needed. The extra stuff you need for shared generics should be encoded first, not last.

**Comment by jkotas** [comment]
cc @JohnChen0

File: src/inc/corinfo.h [hunk 2]
```diff
@@ -636,6 +636,7 @@ enum CorInfoHelpFunc
     CORINFO_HELP_READYTORUN_CHKCAST,
     CORINFO_HELP_READYTORUN_STATIC_BASE,
     CORINFO_HELP_READYTORUN_VIRTUAL_FUNC_PTR,
+    CORINFO_HELP_READYTORUN_GENERIC_DICTIONARY,
```
Comment (jkotas) [inline src/inc/corinfo.h]: I think this should be called `CORINFO_HELP_READYTORUN_GENERIC_HANDLE`.

File: src/inc/corinfo.h [hunk 3]
```diff
@@ -636,6 +636,7 @@ enum CorInfoHelpFunc
     CORINFO_HELP_READYTORUN_CHKCAST,
     CORINFO_HELP_READYTORUN_STATIC_BASE,
     CORINFO_HELP_READYTORUN_VIRTUAL_FUNC_PTR,
```
Comment (jkotas) [inline src/inc/corinfo.h]: Any changes in this file should be under `#if COR_JIT_EE_VERSION > 460` ifdef.

Comment (jkotas) [inline src/inc/corinfo.h]: (Also, there is a special checkin process for them - coordinate with the JIT folks.)

File: src/inc/corinfo.h [hunk 4]
```diff
@@ -2386,6 +2387,7 @@ class ICorStaticInfo
 
     virtual void getReadyToRunHelper(
             CORINFO_RESOLVED_TOKEN * pResolvedToken,
+            CORINFO_LOOKUP_KIND *    pLookupKind,
```
Comment (jkotas) [inline src/inc/corinfo.h]: These should be just one out argument that describes the lookup. The existing `CORINFO_CONST_LOOKUP` argument should be changed to contain what you need - changing the type of the argument to `CORINFO_LOOKUP` may be best.

File: src/vm/prestub.cpp [hunk 5]
```diff
@@ -2359,6 +2360,23 @@ PCODE DynamicHelperFixup(TransitionBlock * pTransitionBlock, TADDR * pCell, DWOR
...
+        {
+            MethodDesc* pContextMD = ZapSig::DecodeMethod(pModule, pInfoModule, pBlob, NULL, &pBlob);
+
+            CORINFO_RESOLVED_TOKEN resolvedToken;
+            resolvedToken.tokenContext = (CORINFO_CONTEXT_HANDLE)pContextMD;
+            resolvedToken.tokenScope = (CORINFO_MODULE_HANDLE)pModule;
+            resolvedToken.token = CorSigUncompressToken(pBlob);
+            resolvedToken.tokenType = (CorInfoTokenKind)CorSigUncompressData(pBlob);
+
+            CEEInfo eeInfo(pContextMD);
```
Comment (jkotas) [inline src/vm/prestub.cpp]: You should fix `ComputeRuntimeLookupForSharedGenericToken` to be a static method and call it directly, without instantiating dummy instance of CEEInfo, to compute the lookup.

File: src/inc/corcompile.h [hunk 6]
```diff
@@ -727,6 +727,9 @@ enum CORCOMPILE_FIXUP_BLOB_KIND
     ENCODE_CHECK_FIELD_OFFSET,
 
     ENCODE_DELEGATE_CTOR,
+
+    ENCODE_GENERIC_DICTIONARY_ACCESS,
```
Comment (jkotas) [inline src/inc/corcompile.h]: The new values in this enum should be prefixes used to describe the kind of the lookup, something like `ENCODE_DICTIONARY_LOOKUP_TYPE` and `ENCODE_DICTIONARY_LOOKUP_METHOD`. They should be followed by regular fixup signature. For example, the signature for `typeof(List<T>)` should be:

`ENCODE_GENERIC_CONTEXT` `ENCODE_TYPE_HANDLE <typespec bytes for List<T> >`

Comment (jkotas) [inline src/inc/corcompile.h]: It may be nice to add `ENCODE_DICTIONARY_LOOKUP_TYPE` and `ENCODE_DICTIONARY_LOOKUP_METHOD` aside from the rest since they are special prefixes - maybe put them at front and give them values 0x8 and 0x9.

Comment (jkotas) [inline src/inc/corcompile.h]: Also, there should be a matching constant `READYTORUN_FIXUP_XXX` added to readytorun.h and `static_assert_no_msg` to src\zap\zapreadytorun.cpp that verify that the two are in sync.

File: src/vm/amd64/cgenamd64.cpp [hunk 7]
```diff
@@ -1126,6 +1126,121 @@ PCODE DynamicHelpers::CreateHelperWithTwoArgs(LoaderAllocator * pAllocator, TADD
...
+    CONTRACTL
+    {
+        GC_NOTRIGGER;
+        PRECONDITION(pCode != NULL && pTarget != NULL);
+    }
+    CONTRACTL_END;
+
+    // Move second argument into argument register (size = 10)
+#ifdef UNIX_AMD64_ABI
+    *(UINT16 *)pCode = 0xBE48; // mov rsi, XXXXXX
```
Comment (jkotas) [inline src/vm/amd64/cgenamd64.cpp]: `CreateHelperWithArg` has exact same code - call this method from there so that it is not duplicated.

File: src/vm/amd64/cgenamd64.cpp [hunk 8]
```diff
@@ -1126,6 +1126,121 @@ PCODE DynamicHelpers::CreateHelperWithTwoArgs(LoaderAllocator * pAllocator, TADD
...
+            *(UINT32*)p = 0x00c98548; p += 3;
+
+            // jne 'POP_AND_RETURN'
+            *(UINT16*)p = 0x1175; p += 2;
+
+            // pop rcx
+            *p++ = 0x59;        // pop the generic context back into rcx
+
+            // mov rdx,pLookup->signature
+            // call pLookup->helper
```
Comment (jkotas) [inline src/vm/amd64/cgenamd64.cpp]: This is `jmp pLookup->helper`

File: src/vm/amd64/cgenamd64.cpp [hunk 9]
```diff
@@ -1126,6 +1126,121 @@ PCODE DynamicHelpers::CreateHelperWithTwoArgs(LoaderAllocator * pAllocator, TADD
...
+            EmitJumpStubWithSecondArg(pAllocator, p, CEEJitInfo::getHelperFtnStatic(pLookup->helper), (TADDR)pLookup->signature);
+
+            // ret
```
Comment (jkotas) [inline src/vm/amd64/cgenamd64.cpp]: The `ret` is not be necessary because the helper was tail called.

File: src/vm/amd64/cgenamd64.cpp [hunk 10]
```diff
@@ -1126,6 +1126,121 @@ PCODE DynamicHelpers::CreateHelperWithTwoArgs(LoaderAllocator * pAllocator, TADD
...
+    else
+    {
+        int codeSize = (pLookup->indirections * 4) + (pLookup->testForNull ? 28 : 4);
+
+        BEGIN_DYNAMIC_HELPER_EMIT(codeSize);
+
+        if (pLookup->testForNull)
+        {
+            // rcx contains the generic context parameter. Push on the stack to save a copy of it.
+            *p++ = 0x51;    // push rcx
```
Comment (jkotas) [inline src/vm/amd64/cgenamd64.cpp]: It would be better to mov into scratch register like rax instead of pushing to stack.

File: src/vm/amd64/cgenamd64.cpp (hunk 8 above)
Comment (fadimounir) [inline src/vm/amd64/cgenamd64.cpp]: Yes. Comment is incorrect. Fixed

---
In reply to: [60820448](https://github.com/dotnet/coreclr/pull/4519#discussion_r60820448) [](ancestors = 60820448)

Comment (fadimounir) [inline src/vm/amd64/cgenamd64.cpp]: @resolved

---
In reply to: [60934450](https://github.com/dotnet/coreclr/pull/4519#discussion_r60934450) [](ancestors = 60934450,60820448)

Comment (fadimounir) [inline src/vm/amd64/cgenamd64.cpp]: Resolved

---
In reply to: [60820448](https://github.com/dotnet/coreclr/pull/4519#discussion_r60820448) [](ancestors = 60820448)

Comment (fadimounir) [inline src/vm/amd64/cgenamd64.cpp]: Resolved

File: src/inc/corinfo.h (hunk 4 above)
Comment (fadimounir) [inline src/inc/corinfo.h]: Changed to just have a CORINFO_LOOKUP argument

---
In reply to: [60819591](https://github.com/dotnet/coreclr/pull/4519#discussion_r60819591) [](ancestors = 60819591)

File: src/inc/readytorun.h [hunk 11]
```diff
@@ -224,6 +224,16 @@ enum ReadyToRunHelper
     READYTORUN_HELPER_Unbox_Nullable            = 0x5B,
     READYTORUN_HELPER_NewMultiDimArr            = 0x5C,
 
+    // Helpers used with generic handle lookup cases
+    READYTORUN_HELPER_NewFast                   = 0x60,
```
Comment (jkotas) [inline src/inc/readytorun.h]: This should be just called NewObject

File: src/jit/gentree.h [hunk 12]
```diff
@@ -2671,7 +2671,7 @@ struct GenTreeCall final : public GenTree
 
 #ifdef FEATURE_READYTORUN_COMPILER
     // Call target lookup info for method call from a Ready To Run module
-    CORINFO_CONST_LOOKUP gtEntryPoint;
+    CORINFO_LOOKUP gtEntryPoint;
```
Comment (jkotas) [inline src/jit/gentree.h]: `CORINFO_LOOKUP` is very large. This will impact JIT memory consumption and throughput. It needs to be done differently.

File: src/zap/zapimport.cpp [hunk 13]
```diff
@@ -1773,6 +1774,45 @@ void ZapImportSectionSignatures::PlaceDynamicHelperCell(ZapImport * pImport)
...
+
+        // token
+        // TODO: stop writing out token, and encode typespec/methodspec/hClass/hField/hMethod members
+        sigBuilder.AppendData(RidFromToken(pResolvedToken->token));
+        sigBuilder.AppendData(TypeFromToken(pResolvedToken->token) >> 24);
+
+        // token type
+        sigBuilder.AppendData(pResolvedToken->tokenType);
+
+        // Dictionary entry kind
```
Comment (jkotas) [inline src/zap/zapimport.cpp]: The encoding of dictionary entry kind should use values from the existing ReadyToRunFixupKind instead to keep the format sane.

Comment (jkotas) [inline src/zap/zapimport.cpp]: (BTW: All enums that can be written into R2R image needs to be captured in readytorun.h.)

File: src/vm/zapsig.cpp (hunk 1 above)
Comment (fadimounir) [inline src/vm/zapsig.cpp]: Removed.

---
In reply to: [60804558](https://github.com/dotnet/coreclr/pull/4519#discussion_r60804558) [](ancestors = 60804558)

File: src/jit/gentree.h (hunk 12 above)
Comment (LLITCHEV) [inline src/jit/gentree.h]: It is much bigger than the CORINFO_CONST_LOOKUP but this is only added for READYTORUN_COMPILER feature defined. As far as I know this is mostly for crossgen a nd never used at runtime jitting. Is that right?

Comment (jkotas) [inline src/jit/gentree.h]: Not anymore ... there is a single JIT .dll for both runtime JITing, crossgen and everything else now.

Comment (fadimounir) [inline src/jit/gentree.h]: I'm going to revert this change, and just pass whatever I need to the APIs separately

Comment (LLITCHEV) [inline src/jit/gentree.h]: Ah, this means that FEATURE_READYTORUN_COMPILER is defined for the runtime JITing compiler too? Is that right? If that is the case, the overhead at runtime will be significant.

**Comment by LLITCHEV** [comment]
Also just a general feedback. Per Jit CC a line should not be longer that 120 characters. Parameters to a call should be all on a single line or one per line, properly aligned.

**Comment by LLITCHEV** [comment]
@pgavlin @russellhadley I know you guys are working on the JIT-EE interface very actively. This PR contains addition of few methods to the jitinterface. Could you, please, coordinate with Fadi on the best way to get this in? Thanks!

**Comment by fadimounir** [comment]
@dotnetbot test this please

**Comment by pgavlin** [comment]
@LLITCHEV @fadimounir sure thing; I'll take a look at this soon.

File: src/jit/gentree.h (hunk 12 above)
Comment (fadimounir) [inline src/jit/gentree.h]: Reverted the change

---
In reply to: [61362958](https://github.com/dotnet/coreclr/pull/4519#discussion_r61362958) [](ancestors = 61362958)

File: src/jit/flowgraph.cpp [hunk 14]
```diff
@@ -7031,8 +7031,7 @@ GenTreePtr    Compiler::fgOptimizeDelegateConstructor(GenTreePtr call, CORINFO_C
 #if COR_JIT_EE_VERSION > 460
                 info.compCompHnd->getReadyToRunDelegateCtorHelper(targetMethod->gtFptrVal.gtLdftnResolvedToken, clsHnd, &call->gtCall.gtEntryPoint);
 #else
-                info.compCompHnd->getReadyToRunHelper(targetMethod->gtFptrVal.gtLdftnResolvedToken,
-                    CORINFO_HELP_READYTORUN_DELEGATE_CTOR, &call->gtCall.gtEntryPoint);
+                info.compCompHnd->getReadyToRunHelper(targetMethod->gtFptrVal.gtLdftnResolvedToken, CORINFO_HELP_READYTORUN_DELEGATE_CTOR, &call->gtCall.gtEntryPoint);
 #endif
```
Comment (LLITCHEV) [inline src/jit/flowgraph.cpp]: This change is probably not needed now.

File: src/jit/importer.cpp [hunk 15]
```diff
@@ -1708,10 +1703,11 @@ GenTreePtr Compiler::impMethodPointer(CORINFO_RESOLVED_TOKEN * pResolvedToken, C
           to lookup the handle.
  */
 
-GenTreePtr          Compiler::impRuntimeLookupToTree(CORINFO_RUNTIME_LOOKUP_KIND kind,
-                                                     CORINFO_RUNTIME_LOOKUP* pLookup,
-                                                     void* compileTimeHandle)
+GenTreePtr          Compiler::impRuntimeLookupToTree(CORINFO_RESOLVED_TOKEN *pResolvedToken, CORINFO_LOOKUP *pLookup, void* compileTimeHandle)
 {
```
Comment (LLITCHEV) [inline src/jit/importer.cpp]: Jit CC guidance - no more than 120 characters per line.

File: src/jit/importer.cpp [hunk 16]
```diff
@@ -1740,47 +1736,55 @@ GenTreePtr          Compiler::impRuntimeLookupToTree(CORINFO_RUNTIME_LOOKUP_KIND
         ctxTree = gtNewLclvNode(info.compTypeCtxtArg, TYP_I_IMPL); // Exact method descriptor as passed in as last arg
     }
 
+#ifdef FEATURE_READYTORUN_COMPILER
+    if (opts.IsReadyToRun())
+    {
+        return impReadyToRunHelperToTree(pResolvedToken, CORINFO_HELP_READYTORUN_GENERIC_HANDLE, TYP_I_IMPL, gtNewArgList(ctxTree), &pLookup->lookupKind);
+    }
```
Comment (LLITCHEV) [inline src/jit/importer.cpp]: Jit CC guidance - no more than 120 characters per line. I think there are more of these below. Please address them too.

File: src/jit/flowgraph.cpp (hunk 14 above)
Comment (fadimounir) [inline src/jit/flowgraph.cpp]: Undone.

---
In reply to: [62374401](https://github.com/dotnet/coreclr/pull/4519#discussion_r62374401) [](ancestors = 62374401)

**Comment by fadimounir** [comment]
@dotnet/jit-contrib PTAL at the JIT changes

**Comment by fadimounir** [comment]
@jo is added to the review. #Closed

**Comment by fadimounir** [comment]
@JohnChen0 is added to the review. #Closed

**Comment by fadimounir** [comment]
Testing work in progress (before I merge anything):
- x86
- x64 on Linux

**Comment by fadimounir** [comment]
@dotnet-bot test OSX x64 Checked Build and Test  please

**Comment by fadimounir** [comment]
@BruceForstall PTAL at the JIT changes

File: src/inc/corinfo.h [hunk 17]
```diff
@@ -1312,6 +1312,7 @@ enum CORINFO_RUNTIME_LOOKUP_KIND
 struct CORINFO_LOOKUP_KIND
 {
     bool                        needsRuntimeLookup;
+    WORD                        runtimeLookupFlags;
```
Comment (pgavlin) [inline src/inc/corinfo.h]: What is the range of values for this field? Is it ever zero if `needsRuntimeLookup` is true?

File: src/jit/importer.cpp [hunk 18]
```diff
@@ -1740,47 +1741,57 @@ GenTreePtr          Compiler::impRuntimeLookupToTree(CORINFO_RUNTIME_LOOKUP_KIND
         ctxTree = gtNewLclvNode(info.compTypeCtxtArg, TYP_I_IMPL); // Exact method descriptor as passed in as last arg
     }
 
+#ifdef FEATURE_READYTORUN_COMPILER
```
Comment (pgavlin) [inline src/jit/importer.cpp]: I'd prefer that we remove these `#ifdef`s and simply rely on `opts.IsReadyToRun`.

Comment (pgavlin) [inline src/jit/importer.cpp]: They are, however, consistent with the rest of the code. We can leave them for now and clean them up later.

**Comment by pgavlin** [comment]
JIT changes LGTM.

File: src/inc/corinfo.h (hunk 17 above)
Comment (fadimounir) [inline src/inc/corinfo.h]: The range is 1-8 today. It should not be zero if needsRuntimeLookup is true (i think some other place in the codebase will catch that and assert on debug builds, and crossgen will not crash, but not compile the affected method)

---
In reply to: [62738274](https://github.com/dotnet/coreclr/pull/4519#discussion_r62738274) [](ancestors = 62738274)

**Comment by LLITCHEV** [comment]
Looks fine to me, assuming the testing legs pass.

**Comment by BruceForstall** [comment]
JIT changes LGTM

**Comment by JohnChen0** [comment]
Runtime changes LGTM

File: src/zap/zapimport.cpp [hunk 19]
```diff
@@ -1773,6 +1775,55 @@ void ZapImportSectionSignatures::PlaceDynamicHelperCell(ZapImport * pImport)
     m_pImage->GetImportTable()->PlaceImportBlob(pCell);
 }
 
+ZapImport * ZapImportTable::GetDictionaryLookupCell(CORCOMPILE_FIXUP_BLOB_KIND kind, CORINFO_RESOLVED_TOKEN * pResolvedToken, CORINFO_LOOKUP_KIND * pLookup)
+{
+    _ASSERTE(pLookup->needsRuntimeLookup);
+
+    // TODO: better way of reusing identical R2R cells!
+    static DWORD* counter = 0;
```
Comment (jkotas) [inline src/zap/zapimport.cpp]: You should use GetImportForSignature instead of this.

File: src/zap/zapimport.cpp [hunk 20]
```diff
@@ -1773,6 +1775,55 @@ void ZapImportSectionSignatures::PlaceDynamicHelperCell(ZapImport * pImport)
...
+    if (!pImport->HasBlob())
+    {
+        SigBuilder sigBuilder;
+
+        sigBuilder.AppendData(kind & ~CORINFO_HELP_READYTORUN_ATYPICAL_CALLSITE);
+
+        //
+        // The token kind we are dealing with for the lookup
+        //
+        sigBuilder.AppendData(pResolvedToken->tokenType);
```
Comment (jkotas) [inline src/zap/zapimport.cpp]: The tokenType should not be encoded in the signature.

File: src/jit/importer.cpp [hunk 21]
```diff
@@ -4580,16 +4594,25 @@ void           Compiler::impImportAndPushBox (CORINFO_RESOLVED_TOKEN * pResolved
...
+            // TODO: ReadyToRun: When generic dictionary lookups are necessary, replace the lookup call
+            // and the newfast call with a single call to a dynamic R2R cell that will:
+            //      1) Load the context
+            //      2) Perform the generic dictionary lookup and caching, and generate the appropriate stub
+            //      3) Allocate and return the new object for boxing
+            // Reason: performance (today, we'll always use the slow helper for the R2R generics case)
+
+            CorInfoHelpFunc helper = opts.IsReadyToRun() ?
+                CORINFO_HELP_NEWFAST :
+                info.compCompHnd->getNewHelper(pResolvedToken, info.compMethodHnd);
```
Comment (jkotas) [inline src/jit/importer.cpp]: If you have a choice, it is always better for R2R special casing to be outside the JIT. 

In this case - we should change getNewHelper to return CORINFO_HELP_NEWFAST unconditionally for R2R in `zap\zapinfo.cpp`.

File: src/inc/corinfo.h (hunk 17 above)
Comment (jkotas) [inline src/inc/corinfo.h]: There should be comment about this - that this is just for internal VM/ZAP communication, not to be used by the JIT.

File: src/vm/jitinterface.cpp [hunk 22]
```diff
@@ -2596,39 +2595,27 @@ void CEEInfo::MethodCompileComplete(CORINFO_METHOD_HANDLE methHnd)
...
+                                                CORINFO_METHOD_HANDLE* ppTemplateMethodHandle /* = NULL */)
 {
-    CONTRACTL {
+    CONTRACTL
+    {
         SO_TOLERANT;
         THROWS;
         GC_TRIGGERS;
         MODE_PREEMPTIVE;
     } CONTRACTL_END;
```
Comment (jkotas) [inline src/vm/jitinterface.cpp]: All methods on the JIT-EE interface need to have these `JIT_TO_EE_TRANSITION` to make exception handling work well. And these transitions cannot be nested - if you need to call the method on the interface internally, there needs to be two different methods, e.g. check `getClassAttribs` vs. `getClassAttribsInternal`.

File: src/vm/jitinterface.cpp [hunk 23]
```diff
@@ -2637,32 +2624,39 @@ void CEEInfo::embedGenericHandle(
...
-        pResult->handleType = CORINFO_HANDLETYPE_CLASS;
-
-        pResult->compileTimeHandle = (CORINFO_GENERIC_HANDLE)th.AsPtr();
+        if (pResult)
+        {
+            pResult->handleType = CORINFO_HANDLETYPE_CLASS;
+            pResult->compileTimeHandle = (CORINFO_GENERIC_HANDLE)th.AsPtr();
+        }
 
         if (fEmbedParent && pResolvedToken->hMethod != NULL)
```
Comment (jkotas) [inline src/vm/jitinterface.cpp]: I doubt that this case is version resilient ... maybe better to fail on it for now.

File: src/jit/importer.cpp [hunk 24]
```diff
@@ -12881,7 +12947,9 @@ MATH_MAYBE_CALL_NO_OVF:         ovfl = false;
 
             JITDUMP(" %08X", resolvedToken.token);
 
-            if (!opts.IsReadyToRun())
+            fNeedsRuntimeLookup = info.compCompHnd->computeNeedsRuntimeLookupAndHandleType(&resolvedToken, FALSE);
```
Comment (jkotas) [inline src/jit/importer.cpp]: We should be calling this only in the readytorun specific codepath.

Comment (jkotas) [inline src/jit/importer.cpp]: I am wondering whether this new method on the JIT-EE interface should be needed at all - the pattern seems to be always:

if (opts.IsReadyToRun())
{
     if (!info.compCompHnd->computeNeedsRuntimeLookupAndHandleType
     {
           info.compCompHnd->getReadyToRunHelper
           ...         
           return; 
     }
     else
     {
           generate the code using the type handle, etc.
     }
}

File: src/jit/importer.cpp (hunk 21 above)
Comment (fadimounir) [inline src/jit/importer.cpp]: Yes, this sounds like a much better idea!

---
In reply to: [62789829](https://github.com/dotnet/coreclr/pull/4519#discussion_r62789829) [](ancestors = 62789829)

File: src/vm/jitinterface.cpp (hunk 23 above)
Comment (fadimounir) [inline src/vm/jitinterface.cpp]: I have an idea on how to make this version resilient, but will do so in another check-in. This scenario is not yet supported with my current changes (crossgen will not compile the method, and we'll fall back to jitting at runtime)

---
In reply to: [62793187](https://github.com/dotnet/coreclr/pull/4519#discussion_r62793187) [](ancestors = 62793187)

File: src/jit/importer.cpp (hunk 24 above)
Comment (fadimounir) [inline src/jit/importer.cpp]: I changed the architecture a little bit, as we discussed, and made the impReadyToRunHelperToTree API return null if a runtime lookup is needed

---
In reply to: [62793570](https://github.com/dotnet/coreclr/pull/4519#discussion_r62793570) [](ancestors = 62793570)

Comment (fadimounir) [inline src/jit/importer.cpp]: Changed the architecture to have the impReadyToRunHelperToTree API return NULL if a runtime lookup is needed

---
In reply to: [62793738](https://github.com/dotnet/coreclr/pull/4519#discussion_r62793738) [](ancestors = 62793738)

File: src/vm/jitinterface.cpp (hunk 22 above)
Comment (fadimounir) [inline src/vm/jitinterface.cpp]: Reverted the addition of the computeNeedsRuntimeLookupAndHandleType

---
In reply to: [62793078](https://github.com/dotnet/coreclr/pull/4519#discussion_r62793078) [](ancestors = 62793078)

File: src/zap/zapimport.cpp [hunk 25]
```diff
@@ -1773,6 +1775,40 @@ void ZapImportSectionSignatures::PlaceDynamicHelperCell(ZapImport * pImport)
...
+    case READYTORUN_FIXUP_TypeHandle:
+        {
+            if (pResolvedToken->pTypeSpec == NULL)
+            {
+                _ASSERTE(!"Invalid IL that directly references __Canon");
+                ThrowHR(E_NOTIMPL);
+            }
+
+            sigBuilder.AppendData(ENCODE_TYPE_HANDLE);
+            sigBuilder.AppendData(pResolvedToken->cbTypeSpec);
```
Comment (jkotas) [inline src/zap/zapimport.cpp]: This should be encoded in the exact same way as the non-generic `ENCODE_TYPE_HANDLE`. There is no number of bytes in the non-generic `ENCODE_TYPE_HANDLE` - it is immediately followed by the typespec.

File: src/zap/zapinfo.cpp [hunk 26]
```diff
@@ -3356,6 +3360,10 @@ unsigned ZapInfo::getClassNumInstanceFields(CORINFO_CLASS_HANDLE cls)
 CorInfoHelpFunc ZapInfo::getNewHelper(CORINFO_RESOLVED_TOKEN * pResolvedToken, CORINFO_METHOD_HANDLE callerHandle)
 {
     classMustBeLoadedBeforeCodeIsRun(pResolvedToken->hClass);
+
+    if (IsReadyToRunCompilation())
```
Comment (jkotas) [inline src/zap/zapinfo.cpp]: The early out for R2R should be the very first thing - before classMustBeLoadedBeforeCodeIsRun.

File: src/vm/amd64/cgenamd64.cpp [hunk 27]
```diff
@@ -1126,6 +1140,97 @@ PCODE DynamicHelpers::CreateHelperWithTwoArgs(LoaderAllocator * pAllocator, TADD
...
+            // rcx contains the value of the dictionary slot entry
+
+            _ASSERTE(pLookup->indirections != 0);
+
+            // test rcx,rcx
+            *(UINT32*)p = 0x00c98548; p += 3;
+
+            // jne 'RETURN' (a jump of 0x12 bytes)
+            *(UINT16*)p = 0x1275; p += 2;
+
```
Comment (jkotas) [inline src/vm/amd64/cgenamd64.cpp]: Forward branches are predicted not taken by the processor by default. It would be better to make this "je" (and swap the blocks below).

File: src/vm/jitinterface.cpp [hunk 28]
```diff
@@ -3187,72 +3162,199 @@ void CEEInfo::ComputeRuntimeLookupForSharedGenericToken(DictionaryEntryKind entr
...
         // If we've got an object, go through its vtable
-        else 
+        else
         {
             _ASSERTE(pContextMD->AcquiresInstMethodTableFromThis());
             pResultLookup->lookupKind.runtimeLookupKind = CORINFO_LOOKUP_THISOBJ;
             pResult->helper = fInstrument ? CORINFO_HELP_RUNTIMEHANDLE_CLASS_LOG : CORINFO_HELP_RUNTIMEHANDLE_CLASS;
         }
+    }
+
```
Comment (jkotas) [inline src/vm/jitinterface.cpp]: I think this can stay inside ComputeRuntimeLookupForSharedGenericTokenStatic to reduce number of arguments.

File: src/vm/prestub.cpp [hunk 29]
```diff
@@ -2338,27 +2340,108 @@ PCODE DynamicHelperFixup(TransitionBlock * pTransitionBlock, TADDR * pCell, DWOR
...
+        
+            DictionaryEntryKind entryKind = EmptySlot;
+            CORCOMPILE_FIXUP_BLOB_KIND signatureKind = (CORCOMPILE_FIXUP_BLOB_KIND)CorSigUncompressData(pBlob);
+
+            switch (signatureKind)
+            {
+            case ENCODE_TYPE_HANDLE:
+                {
+                    entryKind = TypeHandleSlot;
+                    resolvedToken.cbTypeSpec = CorSigUncompressData(pBlob);
```
Comment (jkotas) [inline src/vm/prestub.cpp]: Pass in -1 instead of encoding the number of bytes.

File: src/vm/jitinterface.cpp (hunk 28 above)
Comment (jkotas) [inline src/vm/jitinterface.cpp]: (Pass in pContextMT/pContextMD instead - it is available at the other callsite too.)

---
In reply to: [63012927](https://github.com/dotnet/coreclr/pull/4519#discussion_r63012927) [](ancestors = 63012927)

**Comment by jkotas** [comment]
LGTM modulo comments.

File: src/vm/jitinterface.cpp (hunk 28 above)
Comment (fadimounir) [inline src/vm/jitinterface.cpp]: I'll revisit this in the near future when re-enabling the dictionary lookup optimization for generic arg type handle lookups

---
In reply to: [63013661](https://github.com/dotnet/coreclr/pull/4519#discussion_r63013661) [](ancestors = 63013661,63012927)

**Comment by fadimounir** [comment]
@dotnet-bot test Windows_NT x86 ryujit Checked Build and Test please

**Comment by fadimounir** [comment]
@dotnet-bot test Windows_NT x64 Release Priority 1 Build and Test please

**Comment by kyulee1** [comment]
As you probably got automatic email, this merge caused build break for arm64.
http://dotnet-ci.cloudapp.net/job/dotnet_coreclr/job/master/job/arm64_cross_debug_windows_nt/

**Comment by gkhanna79** [comment]
@fadimounir Can you PTAL?
